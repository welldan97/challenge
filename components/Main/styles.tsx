import styled from 'styled-components';

// SECTION: Main

export const Box = styled.div`
  display: flex;
  flex-direction: column;
  width: ${({ isShrunk = false }: { isShrunk?: boolean }) =>
    isShrunk ? '406px' : '520px'};
  border: 1px solid rgba(105, 50, 212, 0.2);
  border-radius: 8px;
  box-shadow: 0px 8px 24px rgba(32, 37, 50, 0.15);
  height: 446px;
  background: #f5f5ff;
  overflow: hidden;
`;

export const Header = styled.header`
  display: flex;
  align-items: center;
  height: 72px;
  padding-top: 4px;
  padding-left: 24px;

  color: #202532;
  background-color: white;
  border-bottom: 1px solid #e0e0f4;

  font-size: 20px;
`;

export const Footer = styled.footer`
  display: flex;
  align-items: center;
  height: 72px;
  padding-left: 24px;
  padding-right: 24px;

  ${({ isSeamless = false }: { isSeamless?: boolean }) =>
    isSeamless
      ? 'justify-content: end;'
      : `
          justify-content: space-between;
          background-color: white;
          border-top: 1px solid #e0e0f4;
        `}
`;

export const Button = styled.button`
  background: #6932d4;
  border-radius: 100px;
  color: white;
  border: none;
  height: 40px;
  width: 88px;
`;
