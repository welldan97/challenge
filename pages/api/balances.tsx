import { NextApiRequest, NextApiResponse } from 'next';
import { Balance } from '../../lib/Balance';
import db from '../../lib/db';
import { Response } from '../../lib/Response';

// SECTION: Main

export default function handler(
  req: NextApiRequest,
  res: NextApiResponse<Response<Balance[]>>,
) {
  if (req.method !== 'GET') {
    res.status(404).json({ status: 'error', message: 'Not found' });
    return;
  }

  res.status(200).json({
    status: 'success',
    value: db.get(),
  });
}
