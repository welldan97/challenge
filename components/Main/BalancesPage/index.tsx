import React, { memo } from 'react';
import BalancesTable from './BalancesTable';
import { Box, Header, Footer, Button } from '../styles';
import type { Balance } from '../../../lib/Balance';

// SECTION: Main

interface Props {
  onNext: () => void;
  balances: Balance[];
}

const BalancePage = memo(({ onNext, balances }: Props) => (
  <Box>
    <Header>My Ethereum addresses</Header>
    <BalancesTable balances={balances} />
    <Footer>
      <span>Please copy the address from which you wish to send money.</span>
      <Button type="button" autoFocus onClick={onNext}>
        Next
      </Button>
    </Footer>
  </Box>
));

// /SECTION

export default BalancePage;
