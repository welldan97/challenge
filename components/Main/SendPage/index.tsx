import { Formik } from 'formik';
import { memo, useCallback, useMemo } from 'react';
import styled from 'styled-components';
import Input from '../../lib/Input';
import { Box, Button, Footer, Header } from '../styles';
import { Transaction } from '../../../lib/Transaction';

// SECTION: Styles

const Form = styled.form`
  margin-top: 29px;
  margin-bottom: 7px;
`;

// !SECTION

// SECTION: Main

interface SendProps {
  onSend: () => void;
}

const SendPage = ({ onSend }: SendProps) => (
  <Box isShrunk>
    <Header>Please fill the form to send Ethereum</Header>
    <Form onSubmit={onSend} id="transaction">
      <Input name="from" placeholder="Your address" label="From" />
      <Input name="to" placeholder="Destination address" label="To" />
      <Input name="amount" placeholder="Ethereum amount" label="Amount" />
    </Form>
    <Footer isSeamless>
      <Button type="submit" autoFocus onClick={onSend} form="transaction">
        Send
      </Button>
    </Footer>
  </Box>
);

// /SECTION
// SECTION: Container

interface Props {
  onSend: (transaction: Transaction) => void;
}

export default memo(({ onSend }: Props) => {
  const transaction = useMemo(
    () => ({
      from: '',
      to: '',
      amount: '',
    }),
    [],
  );

  return (
    <Formik initialValues={transaction} enableReinitialize onSubmit={onSend}>
      {({ handleSubmit }) => <SendPage onSend={handleSubmit} />}
    </Formik>
  );
});
