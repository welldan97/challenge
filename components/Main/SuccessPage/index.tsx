import { memo } from 'react';
import styled from 'styled-components';
import { Transaction } from '../../../lib/Transaction';
import Image from '../../lib/Image';
import { Box, Header } from '../styles';

// SECTION: Styles

const Success = styled(Image)`
  margin: 27px auto 11px;
  width: 100px;
`;

const Wrapper = styled.div`
  padding-left: 24px;
  padding-right: 24px;
`;

const Label = styled.div`
  font-size: 16px;
`;

const Amount = styled.div`
  font-size: 28px;
  transform: translateY(-2px);
`;

const Hr = styled.hr`
  border-top: 1px solid #202532;
  opacity: 1;
  margin-top: 12px;
  margin-bottom: 18px;
  width: 142px;
`;

const Address = styled.div`
  transform: translateY(1px);
  margin-bottom: 14px;
`;
// !SECTION

// SECTION: Main

interface Props {
  transaction: Transaction;
}

const SuccessPage = ({ transaction }: Props) => (
  <Box isShrunk>
    <Header>Transaction complete</Header>
    <Wrapper>
      <Success src="/Success.svg" alt="success" width="100" height="100" />
      <Label> You sent</Label>
      <Amount>{transaction.amount} ETH</Amount>
      <Hr />
      <Label>From</Label>
      <Address>{transaction.from}</Address>
      <Label>To</Label>
      <Address>{transaction.to}</Address>
    </Wrapper>
  </Box>
);

// /SECTION
// SECTION: Container

export default memo(SuccessPage);
