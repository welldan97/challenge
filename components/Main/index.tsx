import React, { memo, useCallback, useState } from 'react';
import styled from 'styled-components';
import useBalances from '../../api/useBalances';
import type { Balance } from '../../lib/Balance';
import { Transaction } from '../../lib/Transaction';
import Image from '../lib/Image';
import BalancePage from './BalancesPage';
import SendPage from './SendPage';
import SuccessPage from './SuccessPage';

// SECTION: Styles

const Logo = styled(Image)`
  margin-bottom: 21px;
  margin-left: 25px;
`;

// !SECTION
// SECTION: Main

type Page = 'balances' | 'send' | 'success';

interface Props {
  page: Page;
  balances: Balance[];
  transaction?: Transaction;
  onNext: () => void;
  onSend: (transaction: Transaction) => void;
}

const Main = memo(({ page, balances, transaction, onNext, onSend }: Props) => (
  <div>
    <Logo src="/Logo.svg" alt="Utrust" width="120" height="28" />
    {page === 'balances' && <BalancePage onNext={onNext} balances={balances} />}
    {page === 'send' && <SendPage onSend={onSend} />}
    {page === 'success' && <SuccessPage transaction={transaction!} />}
  </div>
));

// !SECTION
// SECTION: Container

export default memo(() => {
  const [page, setPage] = useState<Page>('balances');

  const handleNext = useCallback(() => {
    setPage('send');
  }, []);

  const handleSendSuccess = useCallback(() => {
    setPage('success');
  }, []);

  const { isError, balances, transaction, onSend } = useBalances({
    onSendSuccess: handleSendSuccess,
  });

  if (isError) return null;
  return (
    <Main
      page={page}
      balances={balances!}
      transaction={transaction}
      onNext={handleNext}
      onSend={onSend}
    />
  );
});
