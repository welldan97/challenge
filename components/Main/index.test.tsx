import { render, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import balances from '../../lib/test/fixtures/balances';
import transaction from '../../lib/test/fixtures/transaction';
import Wrapper from '../../lib/test/Wrapper';
import Main from '.';

// SECTION: Main

describe('Main', () => {
  let mock: MockAdapter;

  beforeEach(() => {
    mock = new MockAdapter(axios);

    mock.onGet('/api/balances').reply(200, {
      success: true,
      value: balances,
    });

    mock.onPost('/api/send').reply(200, {
      success: true,
      value: transaction,
    });
  });

  afterEach(() => {
    mock.reset();
  });

  const balancesPage = async () => {
    const { container } = await render(
      <Wrapper>
        <Main />
      </Wrapper>,
    );

    await waitFor(() => {
      expect(screen.getByTestId('wrapper')).toBeTruthy();
    });

    return { container };
  };

  const sendPage = async () => {
    const { container } = await balancesPage();

    const nextButton = screen.getByTestId('next');
    await userEvent.click(nextButton);

    return { container };
  };

  test('balances page', async () => {
    const { container } = await balancesPage();

    expect(container).toMatchSnapshot();
  });

  test('send page', async () => {
    const { container } = await sendPage();

    expect(container).toMatchSnapshot();
  });

  test('success page', async () => {
    const { container } = await sendPage();

    const fromInput = screen.getByTestId('from');
    await userEvent.type(fromInput, 'FROM');

    const toInput = screen.getByTestId('to');
    await userEvent.type(toInput, 'TO');

    const amountInput = screen.getByTestId('amount');
    await userEvent.type(amountInput, 'AMOUNT');

    const sendButton = screen.getByTestId('send');
    await userEvent.click(sendButton);

    expect(container).toMatchSnapshot();
  });
});
