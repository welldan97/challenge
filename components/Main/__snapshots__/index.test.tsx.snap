// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main balances page 1`] = `
<div>
  <div
    data-testid="wrapper"
  >
    <div
      class="sc-60f604f0-0 hxgFlm"
    >
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27120%27%20height=%2728%27/%3e"
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt="Utrust"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </div>
    <div
      class="sc-da1060d0-0 fFYZiQ"
    >
      <header
        class="sc-da1060d0-1 eHCwEr"
      >
        My Ethereum addresses
      </header>
      <table
        class="sc-50244374-0 dsjiEy"
      >
        <tbody>
          <tr
            class="sc-50244374-1 cYIIdr"
          >
            <td>
              <span
                class="sc-50244374-2 bGcnud"
              >
                TEST_ADDRESS1
              </span>
            </td>
            <td
              class="sc-50244374-3 jGtUXk"
            >
              <span
                class="sc-50244374-4 jMsZsK"
              >
                TEST_VALUE1
              </span>
              <span
                class="sc-50244374-5 hbdULL"
              >
                 ETH
              </span>
            </td>
          </tr>
          <tr
            class="sc-50244374-1 cYIIdr"
          >
            <td>
              <span
                class="sc-50244374-2 bGcnud"
              >
                TEST_ADDRESS2
              </span>
            </td>
            <td
              class="sc-50244374-3 jGtUXk"
            >
              <span
                class="sc-50244374-4 jMsZsK"
              >
                TEST_VALUE2
              </span>
              <span
                class="sc-50244374-5 hbdULL"
              >
                 ETH
              </span>
            </td>
          </tr>
          <tr
            class="sc-50244374-1 cYIIdr"
          >
            <td>
              <span
                class="sc-50244374-2 bGcnud"
              >
                TEST_ADDRESS3
              </span>
            </td>
            <td
              class="sc-50244374-3 jGtUXk"
            >
              <span
                class="sc-50244374-4 jMsZsK"
              >
                TEST_VALUE3
              </span>
              <span
                class="sc-50244374-5 hbdULL"
              >
                 ETH
              </span>
            </td>
          </tr>
        </tbody>
      </table>
      <footer
        class="sc-da1060d0-2 cunNGN"
      >
        <span>
          Please copy the address from which you wish to send money.
        </span>
        <button
          class="sc-da1060d0-3 fUffHN"
          data-testid="next"
          type="button"
        >
          Next
        </button>
      </footer>
    </div>
  </div>
</div>
`;

exports[`Main send page 1`] = `
<div>
  <div
    data-testid="wrapper"
  >
    <div
      class="sc-60f604f0-0 hxgFlm"
    >
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27120%27%20height=%2728%27/%3e"
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt="Utrust"
          data-nimg="intrinsic"
          decoding="async"
          src="/Logo.svg"
          srcset="/Logo.svg 1x, /Logo.svg 2x"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </div>
    <div
      class="sc-da1060d0-0 dVsQgh"
    >
      <header
        class="sc-da1060d0-1 eHCwEr"
      >
        Please fill the form to send Ethereum
      </header>
      <form
        class="sc-b81309d0-0 ePlrqb"
        id="transaction"
      >
        <div
          class="sc-6f71e095-0 zKiGe"
        >
          <label
            class="sc-6f71e095-1 dpiAKW"
            for="from"
          >
            From
          </label>
          <input
            class="sc-6f71e095-2 mOreu"
            data-testid="from"
            id="from"
            placeholder="Your address"
            value=""
          />
        </div>
        <div
          class="sc-6f71e095-0 zKiGe"
        >
          <label
            class="sc-6f71e095-1 dpiAKW"
            for="to"
          >
            To
          </label>
          <input
            class="sc-6f71e095-2 mOreu"
            data-testid="to"
            id="to"
            placeholder="Destination address"
            value=""
          />
        </div>
        <div
          class="sc-6f71e095-0 zKiGe"
        >
          <label
            class="sc-6f71e095-1 dpiAKW"
            for="amount"
          >
            Amount
          </label>
          <input
            class="sc-6f71e095-2 mOreu"
            data-testid="amount"
            id="amount"
            placeholder="Ethereum amount"
            value=""
          />
        </div>
      </form>
      <footer
        class="sc-da1060d0-2 MrGpr"
      >
        <button
          class="sc-da1060d0-3 fUffHN"
          data-testid="send"
          form="transaction"
          type="submit"
        >
          Send
        </button>
      </footer>
    </div>
  </div>
</div>
`;

exports[`Main success page 1`] = `
<div>
  <div
    data-testid="wrapper"
  >
    <div
      class="sc-60f604f0-0 hxgFlm"
    >
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27120%27%20height=%2728%27/%3e"
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt="Utrust"
          data-nimg="intrinsic"
          decoding="async"
          src="/Logo.svg"
          srcset="/Logo.svg 1x, /Logo.svg 2x"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </div>
    <div
      class="sc-da1060d0-0 dVsQgh"
    >
      <header
        class="sc-da1060d0-1 eHCwEr"
      >
        Transaction complete
      </header>
      <div
        class="sc-70be64b8-1 eYoFSo"
      >
        <div
          class="sc-70be64b8-0 jflKBd"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27100%27%20height=%27100%27/%3e"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="success"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
        <div
          class="sc-70be64b8-2 hIOwaP"
        >
           You sent
        </div>
        <div
          class="sc-70be64b8-3 hkrVfx"
        >
          TEST_VALUE
           ETH
        </div>
        <hr
          class="sc-70be64b8-4 ifepvz"
        />
        <div
          class="sc-70be64b8-2 hIOwaP"
        >
          From
        </div>
        <div
          class="sc-70be64b8-5 jEjKnN"
        >
          TEST_FROM
        </div>
        <div
          class="sc-70be64b8-2 hIOwaP"
        >
          To
        </div>
        <div
          class="sc-70be64b8-5 jEjKnN"
        >
          TEST_TO
        </div>
      </div>
    </div>
  </div>
</div>
`;
